usr/sbin/p11sak/p11sak.c

```diff
@@ -2421,8 +2812,71 @@ static void free_attributes(CK_ATTRIBUTE *attrs, CK_ULONG num_attrs)
     free(attrs);
 }
 
+static bool is_attr_array_attr(CK_ATTRIBUTE *attr)
+{
+    switch (attr->type) {
+    case CKA_WRAP_TEMPLATE:
+    case CKA_UNWRAP_TEMPLATE:
+    case CKA_DERIVE_TEMPLATE:
+        return true;
+
+    default:
+        return false;
+    }
+}
+
+static void free_attr_array_attr(CK_ATTRIBUTE *attr)
+{
+    CK_ULONG i, num;
+    CK_ATTRIBUTE *elem;
+
+    num = attr->ulValueLen / sizeof(CK_ATTRIBUTE);
+    for (i = 0, elem = attr->pValue; elem != NULL && i < num; i++, elem++) {
+        if (elem->pValue != NULL) {
+            if (is_attr_array_attr(elem))
+                free_attr_array_attr(elem);
+            free(elem->pValue);
+            elem->pValue = NULL;
+        }
+    }
+}
+
+static CK_RV alloc_attr_array_attr(CK_ATTRIBUTE *attr, bool *allocated)
+{
+    CK_ULONG i, num;
+    CK_ATTRIBUTE *elem;
+    CK_RV rc;
+
+    *allocated = false;
+
+    num = attr->ulValueLen / sizeof(CK_ATTRIBUTE);
+    for (i = 0, elem = attr->pValue; i < num; i++, elem++) {
+        if (elem->ulValueLen > 0 && elem->pValue == NULL) {
+            elem->pValue = calloc(elem->ulValueLen, 1);
+            if (elem->pValue == NULL) {
+                free_attr_array_attr(attr);
```

Comment:
attr = NULL; after the free? Also below ... ?

Reply:
Why? the function just returns right after the free, so that variable can't be used afterwards, and free_attr_array_attr() sets attr->pValue to NULL once it freed it.